### PR TITLE
Fix 223

### PR DESCRIPTION
**Pull request by clanmills**
This is a replacement for https://github.com/Exiv2/exiv2/pull/223 and discussed in PR#223

This PR will not pass travis because the changes cause the "old" test suite to detect errors.  I intend to include the test files provided by  @mcdamo in the new test suite.  I will also update travis to build and test using the new test suite.

### CODE REVIEW

**Comment by mcdamo** [comment]
Sorry, this is very minor but I noticed the formatting I used is not consistent with how these are written by Fujifilm and others - there should be no spaces around the '+' symbols. 
```
Monochrome+R Filter
Monochrome+G Filter
Monochrome+Ye Filter 
```
I would be grateful if you could amend this in the patch.

**Comment by clanmills** [comment]
Thanks for bring this to my attention.  I've fixed this:
```
719 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ exiv2 -pa --grep "Fujifilm.Color" ~/Downloads/*.jpg 
/Users/rmills/Downloads/Monochrome+g.jpg  Exif.Fujifilm.Color                          Short       1  Monochrome+G Filter
/Users/rmills/Downloads/Monochrome+r.jpg  Exif.Fujifilm.Color                          Short       1  Monochrome+R Filter
/Users/rmills/Downloads/Monochrome+y.jpg  Exif.Fujifilm.Color                          Short       1  Monochrome+Ye Filter
/Users/rmills/Downloads/Monochrome.jpg  Exif.Fujifilm.Color                          Short       1  Monochrome
/Users/rmills/Downloads/Sepia.jpg  Exif.Fujifilm.Color                          Short       1  Sepia
720 rmills@rmillsmbp:~/gnu/github/exiv2/exiv2 $ 
```

Just for reference, the previous behaviour was:
```
722 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ exiv2 -pa --grep "Fujifilm.Color" ~/Downloads/*.jpg 
/Users/rmills/Downloads/Monochrome+g.jpg  Exif.Fujifilm.Color                          Short       1  (771)
/Users/rmills/Downloads/Monochrome+r.jpg  Exif.Fujifilm.Color                          Short       1  (769)
/Users/rmills/Downloads/Monochrome+y.jpg  Exif.Fujifilm.Color                          Short       1  (770)
/Users/rmills/Downloads/Monochrome.jpg  Exif.Fujifilm.Color                          Short       1  None (black & white)
/Users/rmills/Downloads/Sepia.jpg  Exif.Fujifilm.Color                          Short       1  (784)
723 rmills@rmillsmbp:~/gnu/github/clanmills/exiv2 $ 
```

File: include/exiv2/types.hpp [hunk 1]
```diff
@@ -474,10 +474,8 @@ namespace Exiv2 {
         return rc == src + N ? 0 : rc;
     }
 
-    //! Template used in the COUNTOF macro to determine the size of an array
-    template <typename T, int N> char (&sizer(T (&)[N]))[N];
 //! Macro to determine the size of an array
-#define EXV_COUNTOF(a) (sizeof(Exiv2::sizer(a)))
+#define EXV_COUNTOF(a) (sizeof(a)/sizeof(a[0]))
```
Comment (D4N) [inline include/exiv2/types.hpp]: This might not work as arrays can silently decay to pointers. See: http://blog.natekohl.net/making-countof-suck-less/

**Comment by D4N** [comment]
Why have you removed `tags_int.cpp` and moved all the array definitions into the header file? This will cause odr violations if this header is included in different translation units.

Concerning the attribution of the commit for @mcdamo, I can simply cherry-pick his commit and push it into this branch and drop the two commits from you that were equivalent to his.

**Comment by clanmills** [comment]
I removed tags_int.cpp because it's a corpse and not compiled.  And there was also a lot of duplicated headers in tags_int.hpp and tags.hpp.  So, I consolidated the duplicated headers and remove the corpse.  That reduced the size of the code base by 3000 lines.

The sizeof(a)/sizeof(a[0]) is mentioned in the "White Book" and supported by all compilers - even for empty arrays.

**Comment by clanmills** [comment]
Dan: I think we should integrate this.  I believe EXV_COUNTOF is only used the context of EXV_PRINT_TAG on arrays that are static/read-only.  The code's OK.

We can't replace my stuff with the contribution from @mcdamo because he used the `// to silence compiler warning` which I wish to remove. 

When https://github.com/Exiv2/exiv2/issues/220 has been submitted (with my changes to test/Makefile), I'll add a test to respect mcdamo's test files.  That test will download the reference image files on demand.  When I've done this, we can have an interesting discussion about how to implement this into the "new" test harness.

I really like what's been done with the "new" test harness.  I've proposed to never change test/data/bugfixes-test.out and very pleased that you've agreed that this is the correct way forward.  For now, we're stuck with the binary beast.  However, I think we can kill him off totally in v0.27 when we'll only have the "new" test suite.  He's alive and well in SVN for historical purposes.  This is a little messy and a consequence of moving to git.

**Comment by D4N** [comment]
Robin Mills <notifications@github.com> writes:

> Dan: I think we should integrate this.  I believe EXV_COUNTOF is only used the context of EXV_PRINT_TAG on arrays that are static/read-only.  The code's OK.
>

Concerning the EXV_COUNTOF macro, I like your version better in terms of
readability (it is a lot clearer what it does in contrast to the current
template magic), but it has a downside: you can misuse it inside a
function. Consider the following snippet:
~~~~~~~~~~~~~~~~~~~~~~~~~~~
#include <stdio.h>

#define COUNTOF(arr) (sizeof(arr) / sizeof(arr[0]))

void print_countof(int arr[]) {
  printf("in print_countof: COUNTOF(arr) = %d\n", COUNTOF(arr));
}

int arr[] = {1, 2, 3, 4, 5, 6, 7, 8};

int main() {
  printf("COUNTOF(arr) = %d\n", COUNTOF(arr));
  print_countof(arr);
  return 0;
}
~~~~~~~~~~~~~~~~~~~~~~~~~~~
This prints:
COUNTOF(arr) = 8
in print_countof: COUNTOF(arr) = 2

The issue is that once we pass an array into a function, it decays to a
pointer and sizeof(arr) becomes sizeof(T*).

If you take the current macro, then you'll get a compilation error
inside print_countof() because the template arguments cannot be
deduced. Imho that is also the better solution, especially as we are
providing this macro in the globally visible header.


Concerning the changes to tags(_int).cpp/hpp: I do not think we should
put the array definitions into the header file. Everything that has
external linkage should be defined in source files.

The linker does not complain in this case because of the preprocessor
include guard TAG_DEFINITIONS that is now present in tags_int.hpp and
prevents the inclusion of all the problematic definitions unless you
define it explicitly (which is only done in tags.cpp).

I am sorry, but I don't think that we should do that. A header should only
contain declarations and the definitions should be in actual source
files. This patch kind of achieves that, but in an unexpected way: a
future programmer will wonder why they can't get the tag definitions via
the header, then they'll see that TAG_DEFINIONS macro, define it and
they'll get close to 4k lines of linker errors.

Maybe I am missing a crucial insight here? I am not opposing a
restructuring of the source, but I'd like to avoid unexpected surprises.


> We can't replace my stuff with the contribution from @mcdamo because he used the `// to silence compiler warning` which I wish to remove. 
>

Yes we can, we can simply drop the part of the patch and commit as
@mcdamo, like I have done it when you sent me patches. It won't have the
Github GPG signature, but the merge will, so I think that's ok.

> When https://github.com/Exiv2/exiv2/issues/220 has been submitted (with my changes to test/Makefile), I'll add a test to respect mcdamo's test files.  That test will download the reference image files on demand.  When I've done this, we can have an interesting discussion about how to implement this into the "new" test harness.
>
The new test harness is just python, so if you can download your file
via python, you can do it :)

However, downloading something during a test is not generally a good
idea, as this makes tests require an internet connection (OpenSUSE's OBS
and Fedora's Koji usually block the internet connection during builds,
so packages could not run the full test suite).

Would something speak against putting a stripped version of the test
file into the repository (just as we have done it before)?

> I really like what's been done with the "new" test harness.  I've proposed to never change test/data/bugfixes-test.out and very pleased that you've agreed that this is the correct way forward.  For now, we're stuck with the binary beast.  However, I think we can kill him off totally in v0.27 when we'll only have the "new" test suite.  He's alive and well in SVN for historical purposes.  This is a little messy and a consequence of moving to git.
>

Thanks, it's not perfect, but hopefully good enough™ for now. I think
we'll have to kill the binary beast earlier, as it blocks this PR and
#180 (or just drop bugfixes-test.sh from the Makefile).

**Comment by clanmills** [comment]
Gosh, Dan.  You have amazing insight.  Let me revisit the TAG_DEFINITIONS hackery.

I'm working on modifications to the test harness.  Let's discuss that once I have that ready for you to review.

**Pull request closed without merging**